src/compile.rs

```diff
@@ -236,52 +387,118 @@ impl Compiler {
             Ok(())
         }
     }
+}
+
+#[derive(Debug)]
+enum Hole {
+    None,
+    One(InstIdx),
+    Many(Vec<Hole>),
+}
+
+#[derive(Clone, Debug)]
+enum MaybeInst {
+    Compiled(Inst),
+    Uncompiled(InstHole),
```

Comment:
You could remove one layer of mental indirection by removing `InstHole` and expanding its enum here. Since all options other then `Compiled(Inst)` are instructions that miss one or more `goto`s, I think it makes sense to treat them as uniformly as possible. (Of course, you could also move `Split.?` into `InstHole`, but then you need to write more wrapper functions.)